### PR TITLE
buildRustCrate: add heuristic to picking the right binary source files

### PR DESCRIPTION
**Pull request by andir**
###### Motivation for this change

Cargo has a few odd (old) ways of picking source files if the `bin.path`
attribute isn't given in the Cargo.toml. This commit adds support for
some of those. The previous behaviour always defaulted to `src/main.rs`
which was not always the right choice.

Since there is  look-ahead into the unpacked sources before running the
actual builder the path selection logic has to be embedded within the
build script.

`buildRustCrate` currently supports two ways of running building
binaries when processing a crate:

- Explicit definition of all the binaries (& optionally the paths to
their respective `main.rs`) and,
- if not binary was explictly configured all files matching the patterns
  `src/main.rs`, `src/bin/*.rs`.

When the explicit list is given without path information paths are now
being picked from a list of candidates. The first match wins. The order
is the same as within the cargo compatibility code.

If the crate does not provide any libraries the path `src/{bin_name}.rs`
is also considered.

All underscores within the binary names are translated into dashes (`-`)
before the lookups are made. This seems to be a common convention.

cc some of the people that touched the code before me: @shlevy @Mic92  

###### Things done

- [X] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [X] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by andir** [comment]
@GrahamcOfBorg build way-cooler cargo-update cargo-download cargo-vendor cargo-edit carnix

File: pkgs/build-support/rust/build-rust-crate.nix [hunk 1]
```diff
@@ -282,14 +282,48 @@ let makeDeps = dependencies:
...
+             continue
+        fi
+
+        if [[ -z "$BIN_PATH" ]]; then
+          # heuristic to "guess" the correct source file as found in cargo:
+          # https://github.com/rust-lang/cargo/blob/90fc9f620190d5fa3c80b0c8c65a1e1361e6b8ae/src/cargo/util/toml/targets.rs#L308-L325
+
+          # the first two cases are the "new" default IIRC
+          BIN_NAME_=$(echo $BIN_NAME | sed -e 's/_/-/g')
+          FILES="src/bin/$BIN_NAME_.rs src/bin/$BIN_NAME_/main.rs src/bin/main.rs src/main.rs"
```
Comment (Mic92) [inline pkgs/build-support/rust/build-rust-crate.nix]: Should this be an array instead?

File: pkgs/build-support/rust/build-rust-crate.nix [hunk 2]
```diff
@@ -282,14 +282,48 @@ let makeDeps = dependencies:
...
+                    break
+            fi
+          done
+
+          if [[ -z "$BIN_PATH" ]]; then
+            echo "failed to find file for binary target: $BIN_NAME" >&2
+            exit 1
+          fi
+        fi
+        build_bin $BIN_NAME $BIN_PATH
```
Comment (Mic92) [inline pkgs/build-support/rust/build-rust-crate.nix]: Can these variables contain spaces?

Comment (andir) [reply pkgs/build-support/rust/build-rust-crate.nix]: if they do I would guess the system build to fail in random places... I should add quotes there anyway I guess..

File: pkgs/build-support/rust/build-rust-crate.nix (hunk 1 above)
Comment (andir) [reply pkgs/build-support/rust/build-rust-crate.nix]: Probably should be an array instead. I'll dig how to do that with bash.

**Comment by andir** [comment]
I started restructuring the `buildRustCrate` after some feedback from @Mic92. There are also a bunch of rudimentary tests now for all the edge cases of cargo that I encountered so far.

File: pkgs/build-support/rust/build-rust-crate.nix (hunk 2 above)
Comment (andir) [reply pkgs/build-support/rust/build-rust-crate.nix]: the `buildRustCrate` expression support spaces now up to the point where the stdenv builder doesn't handle them well. I could add an extra level of escaping but the benefit is doubtful.

**Comment by andir** [comment]
@GrahamcOfBorg build buildRustCrateTests

**Comment by andir** [comment]
I started playing with this a bit more and it seems that there is still a lot of headroom to improve in regards to what carnix produces and what we are able to properly build.. The `brotli` crate is a great example...

**Review by Mic92 — APPROVED** [review]
Definitely more readable and build with all our current expression.

**Pull request merged**